### ISSUE TITLE
Allow selecting profile to spawn to from a url param

### ISSUE THREAD

**Issue by yuvipanda** [jupyterhub/kubespawner#189]
We don't want to force users to explicitly choose a profile to spawn to sometimes, instead preferring that the profile be selected via URL parameters. 

For example, in a university running multiple courses, each course would want its own docker image & memory requirements. Each of those can be a profile. When the instructors send out links, they can pre-select which profile the user should be spawned into by specifying it in the URL parameter. This way we retain maximum flexibility for the instructors with minimal confusion for the students.

**Comment by yuvipanda** [comment]
This could be done generically in JupyterHub as a way to map URL query parameters to values in the options form. What do you think, @minrk?

**Comment by minrk** [comment]
Makes sense. Essentially allow all the same POST params to also be specified as GET params?

This can be done now by requesting a server with options in an API request, but that gets you into CORS territory.

There are potential security issues with allowing this from a URL since it allows the link author to take actions on behalf of the user without user input, for instance if the deployment has a free docker image field without a whitelist, it would be easy to share a link that launches a notebook *and* a credentials-stealing or crypto mining script. Allowing this should mean that it's *impossible* to submit a form with bad values. Given that, I think we'd have to implement additional CORS protection, only allowing spawn with get params to come from whitelisted sites. Or at least include a confirmation page, so that the user can see the params that have been specified before confirming that they want this to happen.

**Comment by yuvipanda** [comment]
@minrk hmm, we could just make this opt in? In my case I *do* want link authors to take actions on behalf of users without user input. If I've free docker image field without a whitelist that I'm putting credentials into I've already lost, for example.

However, I really only care about this for profiles like in KubeSpawner (#160 and friends), which are pre-configured. Which profiles are available to users is already set by admin. I think this does have fewer security issues?

**Comment by minrk** [comment]
Yeah, picking from a list is unlikely to be an issue, so it's quite doable for *a spawner* to allow this, but hard for JupyterHub in general.

I was thinking of an exact equivalence, allowing get params to be mapped to user_options, which would be cleaner in a lot of ways. But the safety concerns of the very different model (any website can cause you to make one of these requests with arbitrary params with a link) makes me think it should be treated differently, and perhaps all we should do is allow passing GET params down to the spawner to do with as they wish. Then it's up to the deployment/spawner to make a reasonable choice (i.e. only allow GET params for profile selection, not any free fields).

**Comment by yuvipanda** [comment]
@minrk  'passing GET params down to the spawner' sounds like a great idea to me!

**Comment by clkao** [comment]
I was trying to get the GET param to pre-select one of the profiles in options_form. unfortunately the `spawner.handler` attachment in #2036 is only during the spawn call, so params can't be queried while rendering.

**Comment by reevejd** [comment]
Hi @yuvipanda, I'm also interested in selecting a profile to spawn with a GET param. With https://github.com/jupyterhub/jupyterhub/pull/2036 closed, is this now possible with kubespawner? Are there any resources/examples you could point me towards to show me how to achieve what I want? Thank you

**Comment by scivm** [comment]
I see it is possible to subclass kubespawner and pass parameters in the body as user_options when using API.    I wasn't able to get access to url parameters from out of the box so far.  works in progress.  will edit when working....

```
class MyKubeSpawner(KubeSpawner):
    def start(self):
        self.dynamic_profile = self.user_options.get('dynamic_profile', self.dynamic_profile)
        self.image = self.user_options.get('image', self.image)
        ....  (more custom)
        return super().start()
```

calling program:
data = '{"profile": "RStudio", "dynamic_profile", "True", "image": myimage}'
r = requests.post(api_url + '/users/' + name + '/server', data=data, headers={'Authorization': 'token %s' % token, 'dynamic_profile': 'True',})

and then a dynamic profile selection function:

      from kubespawner.spawner import MyKubeSpawner
      c.JupyterHub.spawner_class = MyKubeSpawner
      profiles = [
          {
            'display_name': 'Datascience environment',
            'description': 'Python, R, and Julia',
            'default': 'true',
            'kubespawner_override': {
               'image': 'scivm/jupyter-datascience:latest',
               'cpu_guarantee': 2,
               'cpu_limit': 2,
               'mem_guarantee': '2096M',
               'mem_limit': '8192M',
               'default_url': '/lab'
          }
      ]
      def dynamic_profile_list(spawner, **kwargs):
        if spawner.handler:
           print("In dynamic profile list")
           dynamic_profile = spawner.handler.get_query_argument("dynamic_profile", False)
           if not dynamic_profile:
              return profiles
           else:
              print("Using dynamic profile")
           # create a dynamic profile
           CPU_MAX = 4
           CPU_DEFAULT = .5
           MEM_MAX = 16
           MEM_DEFAULT = "2G"
           image = spawner.handler.get_query_argument("image", "scivm/jupyter-datascience:latest")

           cpu_limit = int(spawner.handler.get_query_argument("cpu_limit", CPU_DEFAULT))
           if cpu_limit > CPU_MAX:
               cpu_limit = CPU_MAX

           mem_limit = spawner.handler.get_query_argument("mem_limit", MEM_DEFAULT)
           mem_number = int(mem_limit[:-1])
           if ((mem_limit.endswith('G') and mem_number > MEM_MAX) or
               (mem_limit.endswith('M') and mem_number > MEM_MAX * 1024) or
               (mem_limit.endswith('K') and mem_number > MEM_MAX * 1024 * 1024) or
               (mem_limit.endswith('B') and mem_number > MEM_MAX * 1024 * 1024 * 1024)):
               mem_limit = str(MEM_MAX) + "G"

           return [
             {
               'display_name': spawner.handler.get_query_argument("display_name", "dynamic profile"),
               'default': True,
               'kubespawner_override': {
                  'image_spec': image,
                  'cpu_limit': cpu_limit,
                  'mem_limit': mem_limit
                }
             }
           ]
      c.KubeSpawner.profile_list = dynamic_profile_list
      c.KubeSpawner.start_timeout = 60 * 5

**Comment by consideRatio** [comment]
Without having read up on the details, as @stv0g said "This would also close the following issue for us: jupyterhub/kubespawner#189", I'll go ahead and close this issue, which then would be resolved by using JupyterHub 1.2 or later that contains this PR: https://github.com/jupyterhub/jupyterhub/pull/3013
